File: modules/wem-api/src/main/java/com/enonic/wem/api/query/aggregation/TermsAggregationQuery.java
package com.enonic.wem.api.query.aggregation;

public class TermsAggregationQuery
    extends AggregationQuery
{
    public final static int TERM_DEFAULT_SIZE = 10;

    private final String fieldName;

    private final int size;

    public TermsAggregationQuery( final Builder builder )
    {
        super( builder.name );
        this.fieldName = builder.fieldName;
        this.size = builder.size;
    }

    public String getFieldName()
    {
        return fieldName;
    }

    public int getSize()
    {
        return size;
    }

    public static class Builder
        extends AggregationQuery.Builder
    {
        private String fieldName;

        public Builder( final String name )
        {
            super( name );
        }

        private int size = TERM_DEFAULT_SIZE;

        public Builder fieldName( final String fieldName )
        {
            this.fieldName = fieldName;
            return this;
        }

        public Builder size( final Integer size )
        {
            this.size = size;
            return this;
        }

        public TermsAggregationQuery build()
        {
            return new TermsAggregationQuery( this );
        }

    }

}

File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/query/aggregation/TermsAggregationQuery.ts
module api.query.aggregation {

    export class TermsAggregationQuery extends AggregationQuery {

        public static TERM_DEFAULT_SIZE: number = 10;

        private fieldName: string;

        private size: number = TermsAggregationQuery.TERM_DEFAULT_SIZE;

        private orderByDirection: string = TermsAggregationOrderDirection.ASC;

        private orderByType: string = TermsAggregationOrderType.TERM;

        toJson(): api.query.aggregation.AggregationQueryTypeWrapperJson {

            var json: api.query.aggregation.TermsAggregationQueryJson = <api.query.aggregation.TermsAggregationQueryJson>super.toAggregationQueryJson();
            json.fieldName = this.getFieldName();
            json.size = this.getSize();
            json.orderByDirection = this.orderByDirection;
            json.orderByType = this.orderByType;

            return <api.query.aggregation.AggregationQueryTypeWrapperJson> {
                TermsAggregationQuery: json
            };
        }

        constructor(name: string) {
            super(name);
        }

        public setFieldName(fieldName: string) {
            this.fieldName = fieldName;
        }

        public getFieldName(): string {
            return this.fieldName;
        }

        public setSize(size: number) {
            this.size = size;
        }

        public getSize(): number {
            return this.size;
        }
    }

    export class TermsAggregationOrderDirection {
        public static ASC: string = "ASC";
        public static DESC: string = "DESC";
    }

    export class TermsAggregationOrderType {
        public static DOC_COUNT: string = "DOC_COUNT";
        public static TERM: string = "TERM";
    }

}
